src/node_api.cc

```diff
@@ -614,6 +603,43 @@ class SetterCallbackWrapper
   const v8::Local<v8::Value>& _value;
 };
 
+v8::Local<v8::ObjectTemplate> EnvObjectTemplate(
+    v8::Isolate* isolate,
+    v8::Persistent<v8::ObjectTemplate>* obj_template,
+    size_t field_count) {
+  v8::Local<v8::ObjectTemplate> return_value;
+
+  if (obj_template->IsEmpty()) {
+    return_value = v8::ObjectTemplate::New(isolate);
+    return_value->SetInternalFieldCount(field_count);
+    obj_template->Reset(isolate, return_value);
+  } else {
+    return_value = v8::Local<v8::ObjectTemplate>::New(isolate, *obj_template);
+  }
+
+  return return_value;
+}
+
+// Create a wrapper object with an internal field holding the payload and a
+// second internal field identifying the owner.
+napi_status EnvWrapperObject(napi_env env,
+                             v8::Persistent<v8::ObjectTemplate>* obj_template,
+                             v8::Local<v8::Value> payload,
+                             const char type_string[],
+                             v8::Local<v8::Object>* result) {
+  auto wrapper_template = EnvObjectTemplate(env->isolate, obj_template, 2);
+  auto maybe_object =
+      wrapper_template->NewInstance(env->isolate->GetCurrentContext());
+  CHECK_MAYBE_EMPTY(env, maybe_object, napi_generic_failure);
+
+  *result = maybe_object.ToLocalChecked();
+  (*result)->SetInternalField(0, payload);
+  (*result)->SetInternalField(1, v8::External::New(env->isolate,
+    reinterpret_cast<void*>(const_cast<char*>(type_string))));
```

Comment:
Also, it would be cool if the internal field names could all be declared as constants at the top of the file, e.g. `constexpr int kPayloadField = 0` and then using that.